Fix MemoryBackingStoreTransaction::abort to restore state correctly

chirags27 opened the pull request:
#### fa9c47e4f4e1b21433aa2d2a7173489facbafde6
<pre>
Fix MemoryBackingStoreTransaction::abort to restore state correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=258701">https://bugs.webkit.org/show_bug.cgi?id=258701</a>
rdar://111503511

Reviewed by Brady Eidson.

This change fixes what happens when we abort a transaction after
attempting to rename an index. Before this change MemoryObjectStore kept
the new name of the index in its internal state which led to crashes.

* LayoutTests/storage/indexeddb/aborted-rename-of-index-crash-expected.txt: Added.
* LayoutTests/storage/indexeddb/aborted-rename-of-index-crash.html: Added.
* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.cpp:
(WebCore::IDBServer::MemoryBackingStoreTransaction::abort):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::renameIndex):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.h:
(WebCore::IDBServer::MemoryObjectStore::info):

Canonical link: <a href="https://commits.webkit.org/265816@main">https://commits.webkit.org/265816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad54f7cadea3952f88e7ebe509dfeb143c0dc825

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11880 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12079 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12453 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13525 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11317 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11899 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12060 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14176 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12044 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12867 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10057 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13947 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10159 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10790 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17919 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10945 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14112 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11355 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9389 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10530 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10669 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2928 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14813 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
```diff
@@ -510,8 +510,9 @@ void MemoryObjectStore::renameIndex(MemoryIndex& index, const String& newName)
+    ASSERT(m_info.infoForExistingIndex(index.info().identifier()));
 
-    m_info.infoForExistingIndex(index.info().name())->rename(newName);
+    m_info.infoForExistingIndex(index.info().identifier())->rename(newName);
```
Are both `ASSERT(m_info.infoForExistingIndex(...));` statements needed, or can Line 512 be deleted?

I'm asking because keeping both doesn't seem to align with the change to Line 514/515 on the surface—I don't know this code very well, so this is just a case of asking, "Why are these two changes not more similar?".

chirags27 replied:
Line 515 is changed mainly for performance, because when looking by name, we go through each item in the map and compare the name, whereas the actual mapping is maintained by index.

As for the assert, I added that mainly to make sure the name and index don't get out of sync. Basically, it's possible that the state we're tracking has the entry with the name, but the index for that is different, in which case we want to flag that. I should probably update the assert to make sure they are equal.

chirags27 committed:
Fix MemoryBackingStoreTransaction::abort to restore state correctly
https://bugs.webkit.org/show_bug.cgi?id=258701
rdar://111503511

Reviewed by Brady Eidson.

This change fixes what happens when we abort a transaction after
attempting to rename an index. Before this change MemoryObjectStore kept
the new name of the index in its internal state which led to crashes.

* LayoutTests/storage/indexeddb/aborted-rename-of-index-crash-expected.txt: Added.
* LayoutTests/storage/indexeddb/aborted-rename-of-index-crash.html: Added.
* Source/WebCore/Modules/indexeddb/server/MemoryBackingStoreTransaction.cpp:
(WebCore::IDBServer::MemoryBackingStoreTransaction::abort):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.cpp:
(WebCore::IDBServer::MemoryObjectStore::renameIndex):
* Source/WebCore/Modules/indexeddb/server/MemoryObjectStore.h:
(WebCore::IDBServer::MemoryObjectStore::info):

Canonical link: https://commits.webkit.org/265816@main
